### ISSUE TITLE
Problem with Video on iOs

### ISSUE THREAD

**Issue by mfossel** [opentok/opentok-ionic-samples#7]
We are using the cordova plugin
`"Cordova-Plugin-OpenTok": "git+https://github.com/opentok/cordova-plugin-opentok.git",`
in an Ionic 2 app. It is working great on Android and it also works on iOs when we connect to a session where the video has already been started. However, when we connect to a session and the video is started our iOs app crashes. Here is our code.

`	this.session = OT.initSession(this.apiKey, sessionId);
			this.session.connect(token, () => {
				this.session.on({
					streamCreated: (event) => {
						this.session.subscribe(event.stream, 'subscriber');
					},
					streamDestroyed: (event) => {
						console.log('Stream ${event.stream.name} ended because 
                                                ${event.reason}');
					}
				});
			});`

Seems to be crashing on that streamCreated event, any idea why this would happen on iOs? 

Thanks in advance, this plugin has been working great, this is the only problem we have encountered.

**Comment by msach22** [comment]
@mfossel If I understand correctly, the app is crashing when you're publishing your video? Did you add the camera and microphone permissions to the Info.plist file of your application?
Could you also let me know which iOS device you're seeing this happen on?

**Comment by mfossel** [comment]
We are only subscribing to the video. We are not publishing with the mobile app, we are publishing from a web app. We are able to watch a video when we connect to a session where the video has already started but when we are connected and a new stream starts up it crashes. 

We have not done the permissions, but as stated we are able to view streams already started.

We are using an iPhone 7.

**Comment by msach22** [comment]
@mfossel Thanks for letting me know. Unfortunately, I'm unable to reproduce the issue on my end. I cloned this repo and commented out the `publisher` code and was able to subscribe to the web publisher without any issues. I also made sure to connect from the phone first and then publish from the web.

If you're not publishing, you won't need the permissions because you're not accessing the device's camera or microphone.

**Comment by msach22** [comment]
I also noticed that in the code sample you provided, you ran `session.on` inside the connect callback. I recommend doing the following:

```
    this.session = OT.initSession(this.apiKey, this.sessionId);

    this.session.on({
      streamCreated: (event) => {
        this.session.subscribe(event.stream, 'subscriber');
      },
      streamDestroyed: (event) => {
        console.log(`Stream ${event.stream.name} ended because ${event.reason}`);        
      }
    });

    this.session.connect(this.token, () => {
    });

```

**Comment by mfossel** [comment]
Ok, we had it that way at first and tried the other way to see if it worked. We will attempt it again with this code. Not really sure what's going on here.

**Comment by msach22** [comment]
@mfossel sounds good! keep me posted :)

**Comment by mfossel** [comment]
Still trying to figure this out. Kind of stumped right now. We added Ionic's error monitoring and we are not getting an error on the crash. The app just seems to close whenever a new stream is started and only on iOs.

**Comment by msach22** [comment]
@mfossel Are you seeing this in a bare minimum Ionic + OpenTok app?

**Comment by mfossel** [comment]
No, we have some other code in there, although not much. I can download the example project and try it on the same device if you think that would be helpful.

**Comment by msach22** [comment]
Yes, please download this sample repo and try reproducing the issue. I tried reproducing the issue, but I was unable to. I think that would be the best way to isolate the issue.

**Comment by mfossel** [comment]
Actually downloading the project and building it for iOs would be a bit of a pain. I am going to just copy the component from the example and put it on a separate page in the app.

**Comment by msach22** [comment]
@mfossel I'm going to close the issue because I'm unable to reproduce it. Please re-open the issue if you're still seeing the crash in the basic-video-chat sample.

**Comment by mfossel** [comment]
Hey just reproduced this by cloning the project.
So I hit start call and connect to the session and then from the browser I start a stream on the same session, app crashes.

If I start stream from browser and then go in app after and start call it works fine.

From iPhone 7

I commented out the publish code because we are not publishing from app

**Comment by msach22** [comment]
@mfossel Can you send me a screenrecording so I can see exactly where in the process the app is crashing? Please start with a fresh clone so we can elimate any outside factors. If possible, please also send the logs that you're seeing when the crash happens

**Comment by mfossel** [comment]
I am using a fresh clone but I commented out the two lines that publish. So basically I hit start call and it connects me to the session fine, but then when I publish from a web app the mobile app crashes. 

If I publish from the web app first and then go into the mobile app and connect to the session I can see the stream fine. 

It seems to be something on the streamCreated event. Sorry, I do not have any logs. I am not able to debug from the iOS app which is what is making this so hard to figure out.

**Comment by mfossel** [comment]
I was able to get a log of the crash by debugging through Xcode. I attached the log, not sure what to make of it. 

[log.txt](https://github.com/opentok/opentok-ionic-samples/files/1704852/log.txt)

**Comment by mfossel** [comment]
We were able to debug in safair but still getting nothing to work with.

We tried to put a breakpoint on 
`this.session.on({ streamCreated: (event) => { this.session.subscribe(event.stream, 'subscriber'); }`
but it crashes right away. When the publisher starts the stream on another app the iOS app just crashes. 

It has to be something to do with this event but not being able to breakpoint or get any information is making it impossible to figure out.

**Comment by mfossel** [comment]
We have pinpointed the problem to the 'streamCreated' event. It's weird but even when we have this code. 
```
this.session.on({ streamCreated: (event) => { 
console.log(event); }
```
the app still crashes whenever a new stream is created. We are thinking of implementing a workaround and handling the stream events with our own sockets. Maybe this helps you figure it out. We will update you on any progress.

**Comment by msach22** [comment]
@mfossel I just cloned this repo and was able to subscribe to a `stream` successfully. Here's a screenshot of the sample app with a publisher and a subscriber: https://cl.ly/0Q3b1x461m1E

Have you inspected the logs for the session using the Inspector tool?
https://tokbox.com/developer/tools/inspector/

When you say the app crashes, do you see an error or the reason why the app is crashing?

**Comment by mfossel** [comment]
@msach22 No error, it just goes straight to the home screen. We tried to breakpoint within the Ionic files but it never hits it, even in the opentok.js we are not able to hit a breakpoint. I put some error logs above, not sure if they are helpful. That is coming from Xcode.

We are able to subscribe to a stream just fine when it has already been started. It's just when we are already in a session and a new stream comes in that it crashes. Are you on iOS? Everything works for us in android.

Yes, we looked in the inspector and didn't have anything out the ordinary.

**Comment by msach22** [comment]
@mfossel Are you running `ionic serve` and `ionic cordova prepare ios` after you set the breakpoints? Yes, it's the iOS simulator.

Can you clarify what you mean when you say? `We are able to subscribe to a stream just fine when it has already been started and it's just when we are already in a session and a new stream comes in that it crashes.`

**Comment by mfossel** [comment]
We are using Safari to breakpoint a build on an actual iPhone. 

We can connect to a session and see the stream from a publisher already publishing from a web app. However, if we come in first and connect to the session on the iPhone and then the web app publishes the app just crashes.

**Comment by msach22** [comment]
@mfossel I'm not sure if it putting the breakpoint in safari will work because the js/html/css are usually packaged as an asset. 

I've tested both methods that you mention and I'm still not seeing the crash.

**Comment by mfossel** [comment]
We are able to breakpoint everything else. Including when it works and connects to a stream the first time. However, it never hits our breakpoint when a new stream comes in, it just goes to desktop.

I'm not sure. Maybe try it with a .ipa on an iPhone? Or could the problem be related to the fact that we are publishing from a web app?

**Comment by msach22** [comment]
Hmm, I'm not sure about that. It should be fine publishing from a webapp b/c that's how I tested it.

**Comment by mfossel** [comment]
Ok are lost at this point, have no idea what's going on. 

We may attempt a workaround because it works when we connect to a session with the stream already going. We only need the app to view a broadcast from a web app, so it's possible that we can handle the event ourselves and avoid the crash when streamCreated comes in.

**Comment by msach22** [comment]
@mfossel  thanks for letting me know. Unfortunately, I'm still unable to reproduce the issue so it could be something with your environment.

**Comment by mark-veenstra** [comment]
@mfossel We have created a full working app on Android and iOS. We are even able to switch between sessions etc. All working fine.
I really think there is something going on in your code or setup why it is crashing on iOS.
What I do know is that you need to make sure you are only connected to 1 session at the time. So if you want to disconnect from one and reconnect to the second this only works if you wait on the disconnected event on the JS side and when you receive this disconnect on the JS side, then reconnect to a new session.

But maybe there is something else going on, on you side. If you can share some more code.

**Comment by mfossel** [comment]
We were able to implement a workaround by handling the broadcast start event ourselves and only connecting to the session when we know that there is a stream already started. It is not perfect but it seems to work for our purposes. I would love to get it working the right way but not having any idea of the problem we had to move on.

I am convinced it is not our code, as we used the sample project and still had the problem. It may be something to do with our setup, hard to say at this point. We had to use the workaround and move on but we will likely revisit this and I will post any updates here.

**Comment by mark-veenstra** [comment]
@mfossel together with @msach22 we are working on updates on the plugin. They will be released soon. But you could test with this branch if it helps: https://github.com/Mediapioniers/cordova-plugin-opentok/tree/all-fixes

Do not use it in production since this one will get removed as soon as all is merged into 3.2.0.
We fixed a lot incompatibilities between the plugin and the official JS SDK. So this branch is more streamlined with it.

**Comment by mfossel** [comment]
Awesome. 
I have been monitoring the updates you both have been making. I tried it last week but it was still crashing. I will try this branch out ASAP and let you know. Thank you!

**Comment by rituparnaGuha** [comment]
I am create a app in ionic 3 
with video calling feature But In ios platform when I try to connect to the session on the iPhone the app just crashes.
I use "ionic cordova plugin add cordova-plugin-opentok". and in index.html I have add opentok.js.and also in plist file I have added 
<key>NSCameraUsageDescription</key> 
<string>The camera is required to publish video</string> 
<key>NSMicrophoneUsageDescription</key> 
<string>The microphone is required to publish audio</string>
can you help me to solve this problem

Thank You

**Comment by nihao89** [comment]
Hi, @msach22 

I try to use ionic to build an audio app. And I use opentok cordova sdk.

My iPhone’s system version is 12.1.4
cordova-plugin-opentok : 3.4.3
@opentok/client: 2.16.1

When I call OT.checkSystemRequirements(), it returns 1. 
I write down OT.upgradeSystemRequirements() in my code, and when I call OT.initSession(this.apiKey, this.sessionId), the screen show me a message as below
![1](https://user-images.githubusercontent.com/50865106/58081391-99d44d00-7be7-11e9-9f5f-4df5390521a6.jpg)

When I call 
this.publisher = OT.initPublisher('publisher', this.publishOptions , function(error){
      alert(JSON.stringify(error))
}) 

the callback function give me a message as below
![2](https://user-images.githubusercontent.com/50865106/58081572-e3249c80-7be7-11e9-9a76-fba696756d84.jpg)

So my question is what should I do can let my phone support opentok. And what are requirements for opentok. 

Thank you in advance
